src/hffs/fs.py

```diff
@@ -244,19 +217,64 @@ def rm(self, path, recursive=False, maxdepth=None, **kwargs):
             commit_message=kwargs.get("commit_message", commit_message),
             commit_description=kwargs.get("commit_description"),
         )
-        self.invalidate_cache(path=huggingface_hub.constants.REPO_TYPES_URL_PREFIXES.get(repo_type, "") + repo_id)
+        self.invalidate_cache(path=path)
 
-    def ls(self, path, detail=True, **kwargs):
+    def ls(self, path, detail=True, refresh=False, **kwargs):
         path = self._strip_protocol(path)
+        if path not in self.dircache or refresh:
+            repo_type, repo_id, _ = self._resolve_repo_id(path)
+            child_dirs = collections.defaultdict(set)
+            repo_type_and_id_prefix = (
+                huggingface_hub.constants.REPO_TYPES_URL_PREFIXES.get(repo_type, "") + repo_id + "/"
+            )
+            num_parts_in_repo_type_and_id_prefix = repo_type_and_id_prefix.count("/")
+            for repo_file in self._iter_tree(path):
+                file_path = repo_type_and_id_prefix + repo_file["path"]
+                child = {
+                    "name": file_path,
+                    "size": repo_file["size"],
+                    "type": "file",
+                    # extra metadata for files
+                    "blob_id": repo_file["oid"],
+                    "lfs": repo_file.get("lfs"),
+                    "last_modified": _get_tz_aware_date_from_iso_date_str(repo_file["lastCommit"]["date"]),
+                }
+                parents = list(PurePosixPath(file_path).parents)[:-num_parts_in_repo_type_and_id_prefix]
+                parent = str(parents[0])
+                self.dircache.setdefault(str(parent), []).append(child)
+                child = {"name": parent, "size": None, "type": "directory"}
+                for parent in parents[1:]:
+                    parent = str(parent)
+                    if child["name"] in child_dirs[parent]:
+                        break
+                    self.dircache.setdefault(parent, []).append(child)
+                    child_dirs[parent].add(child["name"])
+                    child = {"name": parent, "size": None, "type": "directory"}
         out = self._ls_from_cache(path)
-        if not out:
-            self._dircache_from_repo_info(path)
-        out = self._ls_from_cache(path)
-        if out is None:
-            raise FileNotFoundError(path)
-        if detail:
-            return out
-        return [o["name"] for o in out]
+        return out if detail else [o["name"] for o in out]
+
+    def _iter_tree(self, path: str):
+        path = self._strip_protocol(path)
+        repo_type, repo_id, path_in_repo = self._resolve_repo_id(path)
+        headers = self._api._build_hf_headers()
+        revision = self.revision if self.revision is not None else huggingface_hub.constants.DEFAULT_REVISION
+        next_url = (
+            f"{self._api.endpoint}/api/{repo_type}s/{repo_id}/tree/{quote(revision, safe='')}/{self._parent(path_in_repo)}"
```

Comment:
```suggestion
        repo_type_prefix = huggingface_hub.constants.REPO_TYPES_URL_PREFIXES.get(repo_type, "")
        next_url = (
            f"{self._api.endpoint}/api/{repo_type_prefix}{repo_id}/tree/{quote(revision, safe='')}/{self._parent(path_in_repo)}"
```

Reply:
BTW, this could be removed from huggingface_hub. It has been a few months now that `/api/models/...` is valid and it will become the only API at some point in moon-landing.

Reply:
You can just add `models/` to this dict once it's good for you no ?

Reply:
Yes indeed. What I meant is that `f"{self._api.endpoint}/api/{repo_type}s/{repo_id}/tree/{quote(revision, safe='')}/{self._parent(path_in_repo)}"` is enough now. It's simpler to read and wouldn't hide a bug if repo_type is something unknown (if repo_type="sapce" for example, the url will be `.../api/username/repo_name/...` instead of `.../api/sapce/username/repo_name/...` => we wouldn't see why the repo_type is not good in the error message)